File: samplecode/remoteattestation/Enclave/Enclave.cpp
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License..

#include "Enclave.h"

#include <iostream>

using namespace util;
using namespace std;

Enclave* Enclave::instance = NULL;

Enclave::Enclave() {}

Enclave* Enclave::getInstance() {
    if (instance == NULL) {
        instance = new Enclave();
    }

    return instance;
}


Enclave::~Enclave() {
    int ret = -1;

    if (INT_MAX != context) {
        int ret_save = -1;
        ret = enclave_ra_close(enclave_id, &status, context);
        if (SGX_SUCCESS != ret || status) {
            ret = -1;
            Log("Error, call enclave_ra_close fail", log::error);
        } else {
            // enclave_ra_close was successful, let's restore the value that
            // led us to this point in the code.
            ret = ret_save;
        }

        Log("Call enclave_ra_close success");
    }

    sgx_destroy_enclave(enclave_id);
}



sgx_status_t Enclave::createEnclave() {
    sgx_status_t ret;
    int launch_token_update = 0;
    int enclave_lost_retry_time = 1;
    sgx_launch_token_t launch_token = {0};

    memset(&launch_token, 0, sizeof(sgx_launch_token_t));

    do {
        ret = sgx_create_enclave(this->enclave_path,
                                 SGX_DEBUG_FLAG,
                                 &launch_token,
                                 &launch_token_update,
                                 &this->enclave_id, NULL);

        if (SGX_SUCCESS != ret) {
            Log("Error, call sgx_create_enclave fail", log::error);
            print_error_message(ret);
            break;
        } else {
            Log("Call sgx_create_enclave success");

            ret = enclave_init_ra(this->enclave_id,
                                  &this->status,
                                  false,
                                  &this->context);
        }

    } while (SGX_ERROR_ENCLAVE_LOST == ret && enclave_lost_retry_time--);

    if (ret == SGX_SUCCESS)
        Log("Enclave created, ID: %llx", this->enclave_id);


    return ret;
}


sgx_enclave_id_t Enclave::getID() {
    return this->enclave_id;
}

sgx_status_t Enclave::getStatus() {
    return this->status;
}

sgx_ra_context_t Enclave::getContext() {
    return this->context;
}

File: sgx_tstd/src/enclave.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License..

use core::sync::atomic::{AtomicU64, Ordering};
use crate::io;
use crate::path::{Path, PathBuf};
use crate::sync::SgxThreadSpinlock;
use crate::untrusted::fs;
use sgx_trts::enclave;
use sgx_types::*;

pub use sgx_trts::enclave::SgxThreadPolicy;

static LOCK: SgxThreadSpinlock = SgxThreadSpinlock::new();
static mut ENCLAVE_PATH: Option<PathBuf> = None;
static ENCLAVE_ID: AtomicU64 = AtomicU64::new(0);

///
/// get_enclave_base is to get enclave map base address.
///
#[inline]
pub fn get_enclave_base() -> *const u8 {
    enclave::rsgx_get_enclave_base()
}

///
/// get_enclave_size is to get enclave map size.
///
#[inline]
pub fn get_enclave_size() -> usize {
    enclave::rsgx_get_enclave_size()
}

///
/// get_heap_base is to get heap base address.
///
#[inline]
pub fn get_heap_base() -> *const u8 {
    enclave::rsgx_get_heap_base()
}

///
/// get_heap_size is to get heap size.
///
#[inline]
pub fn get_heap_size() -> usize {
    enclave::rsgx_get_heap_size()
}

///
/// get_rsrv_base is to get reserved memory base address.
///
#[inline]
pub fn get_rsrv_base() -> *const u8 {
    enclave::rsgx_get_rsrv_base()
}

///
/// get_rsrv_size is to get reserved memory size.
///
#[inline]
pub fn get_rsrv_size() -> usize {
    enclave::rsgx_get_rsrv_size()
}

///
/// get_tcs_max_num is to get max tcs number.
///
#[inline]
pub fn get_tcs_max_num() -> u32 {
    enclave::rsgx_get_tcs_max_num()
}
///
/// get_thread_policy is to get TCS policy.
///
#[inline]
pub fn get_thread_policy() -> SgxThreadPolicy {
    enclave::rsgx_get_thread_policy()
}

///
/// get_enclave_id is to get enclave ID.
///
pub fn get_enclave_id() -> sgx_enclave_id_t {
    ENCLAVE_ID.load(Ordering::SeqCst) as sgx_enclave_id_t
}

///
/// set_enclave_id is to set enclave ID.
///
pub fn set_enclave_id(eid: sgx_enclave_id_t) {
    ENCLAVE_ID.store(eid, Ordering::SeqCst);
}

///
/// get_enclave_path is to get the path or name of the enclave.
///
pub fn get_enclave_path() -> Option<PathBuf> {
    unsafe {
        LOCK.lock();
        let path = ENCLAVE_PATH.as_ref().map(|p| p.to_owned());
        LOCK.unlock();
        path
    }
}

///
/// set_enclave_path is to set the path or name of the enclave.
///
pub fn set_enclave_path<P: AsRef<Path>>(path: P) -> io::Result<()> {
    let _ = fs::metadata(&path)?;
    unsafe {
        LOCK.lock();
        if ENCLAVE_PATH.is_none() {
            ENCLAVE_PATH = Some(path.as_ref().to_owned());
        }
        LOCK.unlock();
        Ok(())
    }
}